### PR TITLE
BTAT-8058 Adding new view/controller

### PR DESCRIPTION
**Pull request by Tasmine10ten**
Rebased Tuesday - 16.30

### CODE REVIEW

File: app/controllers/contactPreference/EmailPreferenceController.scala [hunk 1]
```diff
@@ -25,22 +25,29 @@ import forms.YesNoForm
...
 
-import scala.concurrent.Future
+import scala.concurrent.{ExecutionContext, Future}
 
-class EmailPreferenceController @Inject()(errorHandler: ErrorHandler,
-                                          emailPreferenceView: EmailPreferenceView)
+class EmailPreferenceController @Inject()(vatSubscriptionService: VatSubscriptionService,
+                                          errorHandler: ErrorHandler,
+                                          emailPreferenceView: EmailPreferenceView,
+                                          addEmailAddressView: AddEmailAddressView)
```
Comment (miranda-hawkes) [inline app/controllers/contactPreference/EmailPreferenceController.scala:39]: This isn't used

File: app/controllers/contactPreference/EmailPreferenceController.scala [hunk 2]
```diff
@@ -54,8 +61,24 @@ class EmailPreferenceController @Inject()(errorHandler: ErrorHandler,
...
+              case Right(details) =>
+               details.ppob.contactDetails.flatMap(_.emailAddress) match {
+                  case Some(_) =>
+                    Redirect(controllers.contactPreference.routes.EmailToUseController.show())
+                      .addingToSession(SessionKeys.contactPrefUpdate -> "true")
+                  case None =>
+                    Redirect(controllers.contactPreference.routes.AddEmailAddressController.show())
+                      .addingToSession(SessionKeys.contactPrefUpdate -> "true")
+
+                }
```
Comment (miranda-hawkes) [inline app/controllers/contactPreference/EmailPreferenceController.scala:75]: You could do something like
```suggestion
val result = details.ppob.contactDetails.flatMap(_.emailAddress) match {
  case Some(_) =>
    Redirect(controllers.contactPreference.routes.EmailToUseController.show())
  case None =>
    Redirect(controllers.contactPreference.routes.AddEmailAddressController.show())
}
result.addingToSession(SessionKeys.contactPrefUpdate -> "true")
```
so you don't have to write the adding to session bit twice

File: it/pages/contactPreference/AddEmailAddressPage.scala [hunk 3]
```diff
@@ -0,0 +1,86 @@
...
+            httpStatus(Status.OK),
+            pageTitle(generateDocumentTitle("cPrefAddEmail.title"))
+          )
+        }
+      }
+    }
+
+    "a user is an authenticated agent" should {
+
+      "render the Agent unauthorised page" in {
```
Comment (miranda-hawkes) [inline it/pages/contactPreference/AddEmailAddressPage.scala:56]:
```suggestion
      "redirect to Client VAT Account" in {
```

File: it/pages/contactPreference/EmailPreferencePageSpec.scala [hunk 4]
```diff
@@ -121,6 +125,52 @@ class EmailPreferencePageSpec extends BasePageISpec {
...
+          When("'Yes' is selected")
+          val res = submit(Yes)
+
+          res should have(
+            httpStatus(Status.SEE_OTHER),
+            redirectURI(controllers.contactPreference.routes.AddEmailAddressController.show().url)
+          )
+        }
+
+        s"add the ${SessionKeys.contactPrefUpdate} key and value to session" in {
```
Comment (miranda-hawkes) [inline it/pages/contactPreference/EmailPreferencePageSpec.scala:146]: I wouldn't add this as a separate test as integration tests are quite slow and it covers the same scenario above. I'd suggest either adding the `SessionCookieCrumbler.getSessionMap(result).get(SessionKeys.contactPrefUpdate) shouldBe Some("true")` check to the above test or just testing the session part in the controller unit test spec

File: test/controllers/contactPreference/AddEmailAddressControllerSpec.scala [hunk 5]
```diff
@@ -0,0 +1,79 @@
...
+
+package controllers.contactPreference
+
+import common.SessionKeys
+import controllers.ControllerBaseSpec
+import play.api.http.Status.{NOT_FOUND, OK}
+import play.api.test.Helpers._
+import views.html.contactPreference.AddEmailAddressView
+
+class AddEmailAddressControllerSpec extends ControllerBaseSpec {
```
Comment (miranda-hawkes) [inline test/controllers/contactPreference/AddEmailAddressControllerSpec.scala:41]: Need to add a scenario for an unauthorised user and a user that currently has a digital preference. I'd also suggest structuring the tests slightly differently:
```
"AddEmailAddressController .show is called" when {

  "user is authorised" when {

    "user has paper preference" when {

      "feature switch is on" should {

      }

      "feature switch is off" should {

      }
    }
  
    "user has a digital preference" should {

    }
  }

  "user is unauthorised" should {
   
  }
}
```
as this reflects the order the logic applies (including the predicates), so its sometimes useful to lay out the tests in the same fashion

File: it/pages/contactPreference/EmailPreferencePageSpec.scala (hunk 4 above)
Comment (Tasmine10ten) [reply it/pages/contactPreference/EmailPreferencePageSpec.scala:146]: Removed as covered in unit tests

File: app/views/contactPreference/AddEmailAddressView.scala.html [hunk 6]
```diff
@@ -0,0 +1,65 @@
...
+@import views.html.templates.inputs.RadioGroup
+@import uk.gov.hmrc.play.views.html.helpers.FormWithCSRF
+@import forms.YesNoForm._
+@import views.html.templates.errors.ErrorSummary
+
+@this(mainTemplate: MainTemplate, radioGroup: RadioGroup, errorSummary: ErrorSummary, formWithCSRF: FormWithCSRF)
+
+@(form: Form[YesNo])(implicit user: User[_], messages: Messages, appConfig: config.AppConfig)
+
+@furtherContent = {
```
[the remark this answers is not in the document]
Comment (Tasmine10ten) [reply app/views/contactPreference/AddEmailAddressView.scala.html:26]: I think this should be resolved now but lemme know if not

**Pull request merged**